README
== Liferay & ActiveRecord

This is a collection of (some) Liferay models for developers that use the database with ActiveRecord. You may want to check out the Rails-portlet package if you're going to deploy Rails applications on Liferay.


== Follow these steps

1. Configure ActiveRecord as usual.
2. require 'lportal'
3. You're good to go.

This gem quite effectively pollutes your namespace, and that is intentional. If you are working in the context of the portal, having your users modeled by User is more convenient rather than having to reference them as "Lportal::Users".


== General notions and examples
=== Caterpillar
To create new instances you need Caterpillar. Liferay's ORM, Hibernate, does not seem to use database sequences, but ActiveRecord assumes they exist. Caterpillar's purpose is to lower the barriers to deploy Rails portlets and it contains various tasks that do not belong to this library. See the Caterpillar docs on installation; this document assumes your environment is set up. Do not forget to run the migrations!


=== Layouts
Each Group can have public and private pages. In both cases, there has to first exist a Web::LayoutSet, which is automatically created for a new Group. The individual pages (class Web::Layout) belong to this set, but they are not directly linked (in 5.1.1 at least). Instead, they belong to the Group and are either public or private. The models have methods to query the relations.

Say you create a new Group "group".

  company = Company.first
  group = Group.create(
    :company => company,
    :creatoruserid => company.administrators.first.id,
    :name => 'Example'
  )

  => #<Group groupid: 8400004, companyid: 10109, creatoruserid: 10129, classnameid: 0, classpk: 0, parentgroupid: 0, livegroupid: 0, name: "Example", description: "", type_: 1, typesettings: "", friendlyurl: "/example", active_: true>

The LayoutSets can then be queried:

  group.public_layoutset

  => #<Web::LayoutSet layoutsetid: 8400011, groupid: 8400004, companyid: 10109, privatelayout: false, logo: false, logoid: 0, themeid: "", colorschemeid: "01", wapthemeid: "mobile", wapcolorschemeid: "01", css: "", pagecount: 0, virtualhost: "">

  group.private_layoutset

  => #<Web::LayoutSet layoutsetid: 8400010, groupid: 8400004, companyid: 10109, privatelayout: true, logo: false, logoid: 0, themeid: "", colorschemeid: "01", wapthemeid: "mobile", wapcolorschemeid: "01", css: "", pagecount: 0, virtualhost: "">

  group.layoutsets.size

  => 2

And a new empty Layout can be created:

  layout = Web::Layout.create(
    :group => group,
    :privatelayout => true,
    :name => 'Private page'
  )

  => #<Web::Layout plid: 8400000, groupid: 8400004, companyid: 10109, privatelayout: true, layoutid: 1, parentlayoutid: 0, name: "<?xml version='1.0' encoding='UTF-8'?><root availab...", title: "<root />", description: "", type_: "portlet", typesettings: "layout-template-id=2_columns_ii\n", hidden_: false, friendlyurl: "/1", iconimage: false, iconimageid: 0, themeid: nil, colorschemeid: nil, wapthemeid: nil, wapcolorschemeid: nil, css: nil, priority: 0, dlfolderid: 0>

You can view this page in the Liferay "Communities" portlet via "Available Communities" => "Actions" => "Manage Pages" => "Private Pages" => "View Pages".


=== Portlets and Caterpillar
To add new portlets the Caterpillar mixins and migrations are required. This data does not live in the lportal database, and this library does not understand about the environment in which Liferay is. Caterpillar, however, does. It parses some Liferay XML configuration files and makes the data available by defining a couple of new methods to classes in the Web module.

The Layout settings are configured via Web::Typesettings. The ”settings” method handles the conversion between the ”typesettings” string and the object model. To add a message_boards portlet into column 2:

  layout.settings = Web::Typesettings.new.message_boards(:column => 2)

  => "layout-template-id=2_columns_ii\ncolumn-2=19,"


The other way around, find the private layouts of "group" that have the Message Board portlet:

  layouts = group.select_layouts_with('message_boards',:private)

Which is a short form for:

  layouts = group.private_layouts.select{|l| l.settings.include?('message_boards')}

  => [#<Web::Layout plid: 8400000, groupid: 8400004, companyid: 10109, privatelayout: true, layoutid: 1, parentlayoutid: 0, name: "<?xml version='1.0' encoding='UTF-8'?><root availab...", title: "<root />", description: "", type_: "portlet", typesettings: "layout-template-id=2_columns_ii\ncolumn-2=19,", hidden_: false, friendlyurl: "/1", iconimage: false, iconimageid: 0, themeid: nil, colorschemeid: nil, wapthemeid: nil, wapcolorschemeid: nil, css: nil, priority: 0, dlfolderid: 0>]

Or query what portlets are in the Layout:

  layouts.first.portlets.each{|p| puts p.title}

  Message Boards
  => [#<Web::PortletPreferences portletpreferencesid: 10210, ownerid: 0, ownertype: 3, plid: 10146, portletid: "19", preferences: "<portlet-preferences xmlns=\"http://java.sun.com/xml...">]

Or whether the layouts include a specific portlet:

  group.layouts_include?('message_boards')
  => true
  group.layouts_include?('message_boards',:private)
  => true
  group.layouts_include?('message_boards',:public)
  => false




Copyright (c) 2008,2009 Mikael Lammentausta, released under the MIT license


ChangeLog
= 1.0.20
	- updated gemfile to include missing required files

= 1.0.19
	- Contry model
	- backports from master branch

= 1.0.18
	- backports from master branch

= 1.0.17
	- introduced schema versions, automatic detection from the release_ table
	- migrations imported from caterpillar trunk, define the migrations directory for caterpillar
	- Group#asset_viewer_portlet(), the asset viewer selection is based on the schema version
	- Liferay portlet path definitions (tagged_content, asset_publisher)
	- Journal::Article#path(), Tag::Asset#path()
	- fixed journal article link
	- Web::PortletPreferences and Web::Portlet: cleanup
	- DRY up various models with Acts::Resourceful
	- api change; layout.contents -> layout.settings
	- api change; user.user_permissions -> user.permissions
	- self.get methods to ResourceCode, Resource and Permission
	- Typesettings: accept Strings and Portlet* models
	- Company: methods to find various groups by type


= 1.0.9
	- creation of MB::Messages
	- new models: SocialActivity, SocialRelation, RatingsStats, MB::MessageFlag, MB::StatsUser
	- URL path to download Document Library File
	- helper methods to find various types of groups and groups assets tags


= 1.0.8
	- Uniqueness scopes in User
	- Group members taken from Organization, if Group type if Organization
	- Role creation and deletion
	- Group creation
	- Added a method to clean Phone number
	- Web::Layout creation
	- Typesettings model for Layout content
	- MB::Category creation, MB::Category.path
	- User.activate, User.portrait_path, User.path
	- Group.name returns the name of Group's owner, if classnameid != 0
	- Layout portlets helper methods
	- fixed Tag::Asset relations

Known bugs:
	- Typesettings does not parse all existing Layouts correctly!


= 1.0.7
	- User creation
	- new models: Listtype, Announcement::Delivery
	- Counter increment method
	- better creation of Phone, Contact and Group


= 1.0.6
	- fixed a bug in user.gsm= method
	- portlet names
	- Company's guest
	- layouts and various Liferay URL handling


= 1.0.5
	- Group types


= 1.0.4
	- Saving user's birthday through User.birthday=
	- New relations
	- New models
	- More tests!


= 1.0.0
Initial release on 18.11.2008.

The structure is usable (though not complete), so tagged as version 1.0.0.



lib/group.rb
# encoding: utf-8
# This model represents the user groups (communities) in the lportal database.
class Group < ActiveRecord::Base
  set_table_name       :group_
  set_primary_key      :groupid

  acts_as_resourceful

  public

  # com.liferay.portal.model.Group
  def liferay_class
    'com.liferay.portal.model.Group'
  end

  # Actions for Permissions.
  def self.actions
    %w{
    APPROVE_PROPOSAL
    ASSIGN_MEMBERS
    ASSIGN_REVIEWER
    DELETE
    MANAGE_ANNOUNCEMENTS
    MANAGE_ARCHIVED_SETUPS
    MANAGE_LAYOUTS
    MANAGE_STAGING
    PERMISSIONS
    PUBLISH_STAGING
    UPDATE
    }
  end

  # Creates a new group (community).
  #
  #
  # Some instances are created based on the Classname of the Group.
  # User's own group, for instance, has to be created with classnameid and classpk.
  #
  # The Group may also have public and private Web::LayoutSet s.
  #
  # This process is engineered by creating a new role with Liferay's (v. 5.1.1) tools and
  # inspecting the database dump diffs.
  #
  # Mandatory parameters:
  #  - companyid
  #  - creatoruserid
  def initialize(params)
    raise 'No companyid given' unless (params[:companyid] or params[:company])
    raise 'No creatoruserid given' unless params[:creatoruserid]
    raise 'No name given' if (
      params[:name].nil? and (params[:classnameid].nil? or params[:classnameid]==0))

    super(params)

    # COPY group_ (groupid, companyid, creatoruserid, classnameid, classpk, parentgroupid, livegroupid, name, description, type_, typesettings, friendlyurl, active_) FROM stdin;
    # +10155	10109	10129	0	0	0	0	testiyhteisö open	testataan yhteisön luomista	1		/10155	t

    self.classnameid   ||= 0 # Community
    self.classpk       ||= 0 # -"-
    self.parentgroupid ||= 0
    self.livegroupid   ||= 0
    self.name          ||= ''
    self.description   ||= ''
    self.type_         ||= 1
    self.typesettings  ||= ''
    self.active_       ||= true

    unless self.friendlyurl
      self.friendlyurl = (
        case self.classnameid
        when 0
          '/'+self.name.downcase
        else
          self.friendlyurl   ||= '/%i' % self.id
        end
      )
    end

    self.save

    # public and private layoutsets

    # COPY layoutset (layoutsetid, groupid, companyid, privatelayout, logo, logoid, themeid, colorschemeid, wapthemeid, wapcolorschemeid, css, pagecount, virtualhost) FROM stdin;
    # +10156	10155	10109	t	f	0		01	mobile	01		0	
    # +10157	10155	10109	f	f	0		01	mobile	01		0	

    [true, false].each do |privacy|
      Web::LayoutSet.create(
        :groupid       => self.id,
        :companyid     => self.companyid,
        :privatelayout => privacy
      )
    end

    ### GROUP
    unless self.classnameid == 0
      # Type is != 0 only for Groups of type Group (?)
      self.type_ = 0
      self.save
    else

      # COPY resource_ (resourceid, codeid, primkey) FROM stdin;
      # +41	6	10155
      # 
      # Create a resource with scope=4 for this Layout.
      r = get_resource(:scope => 4)

      # Permissions (given to administrators)

      # 
      # COPY permission_ (permissionid, companyid, actionid, resourceid) FROM stdin;
      # +81	10109	APPROVE_PROPOSAL	41
      # +82	10109	ASSIGN_MEMBERS	41
      # +83	10109	ASSIGN_REVIEWER	41
      # +84	10109	DELETE	41
      # +85	10109	MANAGE_ANNOUNCEMENTS	41
      # +86	10109	MANAGE_ARCHIVED_SETUPS	41
      # +87	10109	MANAGE_LAYOUTS	41
      # +88	10109	MANAGE_STAGING	41
      # +89	10109	PERMISSIONS	41
      # +90	10109	PUBLISH_STAGING	41
      # +91	10109	UPDATE	41
      # 

      self.class.actions.each do |actionid|
        p = Permission.get(
          :companyid  => self.companyid,
          :actionid   => actionid,
          :resourceid => r.id
        )
        self.company.administrators.each do |user|
          user.permissions << p
        end
      end
    end
  end

  belongs_to :company,
    :foreign_key => 'companyid'

  belongs_to :creator,
    :class_name  => 'User',
    :foreign_key => 'creatoruserid'

  belongs_to :parent,
    :class_name  => 'Group',
    :foreign_key => 'parentgroupid'

  # association to organizations
  has_and_belongs_to_many  :organizations,
                           :join_table              => 'groups_orgs',
                           :foreign_key             => 'groupid',
                           :association_foreign_key => 'organizationid'
  alias :orgs :organizations

  # association to users
  has_and_belongs_to_many  :users,
                           :join_table              => 'users_groups',
                           :foreign_key             => 'groupid',
                           :association_foreign_key => 'userid'

  # association to roles
  has_and_belongs_to_many  :roles,
                           :join_table              => 'groups_roles',
                           :foreign_key             => 'groupid',
                           :association_foreign_key => 'roleid'

  # association to group permissions.
  has_and_belongs_to_many  :permissions,
                           :join_table              => 'groups_permissions',
                           :foreign_key             => 'groupid',
                           :association_foreign_key => 'permissionid'

  # association to group usergroups.
  has_and_belongs_to_many  :usergroups,
                           :join_table              => 'groups_usergroups',
                           :foreign_key             => 'groupid',
                           :association_foreign_key => 'usergroupid'

  ####

  # association to assets
  has_many :assets,
    :class_name  => 'Tag::Asset',
    :foreign_key => 'groupid'

  has_many :mbcategories,
    :class_name => 'MB::Category',
    :foreign_key => 'groupid'

  # a group may have two layoutsets (public and private)
  has_many :layoutsets,
    :class_name  => 'Web::LayoutSet',
    :foreign_key => 'groupid'

  has_one :public_layoutset,
    :class_name  => 'Web::LayoutSet',
    :foreign_key => 'groupid',
    :conditions  => 'privatelayout = false'

  has_one :private_layoutset,
    :class_name  => 'Web::LayoutSet',
    :foreign_key => 'groupid',
    :conditions  => 'privatelayout = true'

  # a group may have many layouts (public and private)
  has_many :layouts,
    :class_name  => 'Web::Layout',
    :foreign_key => 'groupid'

  has_many :public_layouts,
    :class_name  => 'Web::Layout',
    :foreign_key => 'groupid',
    :conditions  => 'privatelayout = false'

  has_many :private_layouts,
    :class_name  => 'Web::Layout',
    :foreign_key => 'groupid',
    :conditions  => 'privatelayout = true'


  # --
  # Class methods, in functional paradigm
  # ++

  class << self

    # 0 = personal user groups
    def find_personal(companyid)
      self.find(:all, :conditions => "companyid=#{companyid} AND type_=0")
    end

    # 1 = public, guest can see
    def find_public(companyid)
      self.find(:all, :conditions => "companyid=#{companyid} AND type_=1")
    end

    # 2 = user-created open communities
    def find_open(companyid)
      self.find(:all, :conditions => "companyid=#{companyid} AND type_=2")
    end

    # 3 = private, also contains admins
    def find_private(companyid)
      self.find(:all, :conditions => "companyid=#{companyid} AND type_=3")
    end
  end


  # --
  # Instance methods
  # ++

  # Group's name. If classnameid==0 (Group is a basic group), use the name in the group_ column.
  # Otherwise use the owner's name.
  def name
    self.classnameid != 0 ?
      self.owner.name : super
  end

  # find owner by classnameid + classpk
  def owner
    return nil if self.classnameid==0
    _class = Classname.model(Classname.find(self.classnameid).value)
    _class.find self.classpk
  end

  # Members by the group type, ie. if the group is of type Organization, return Organization members.
  def members
    if self.owner.kind_of?(Organization)
      return self.owner.members
    else
      #[self.users, self.organizations.collect(&:users)].flatten
      return self.users
    end
  end

  # Assign User to be a member of this Group
  def assign_membership(user)
    raise 'TODO'
    # COPY usergrouprole (userid, groupid, roleid) FROM stdin;
    # +10129	10155	10120
    # +10129	10155	10119
    # 
    # COPY users_groups (userid, groupid) FROM stdin;
    # +10129	10155
    # 
  end


  has_many :mbcategories,
    :class_name  => 'MB::Category',
    :foreign_key => 'groupid'

  def is_active?
    self.active_
  end

  def is_personal?
    self.type_ == 0
  end

  def is_public?
    self.type_ == 1
  end

  def is_protected?
    self.type_ == 2
  end

  def is_private?
    self.type_ == 3
  end

  # URL path to this Group's public or private page
  def path(pl=:public)
    case pl
    when :public
      self.public_layouts.any? ?
        self.public_layoutset.url_prefix + self.friendlyurl : ''

    when :private
      self.private_layouts.any? ?
        self.private_layoutset.url_prefix + self.friendlyurl : ''

    else
      ''
    end
  end

  # Selects the layout (from both public and private) that has the portlet.
  #
  # Parameters:
  #  - portlet     can be either a String of portlet name (eg. 'message_boards'),
  #                or a Web::Portlet, Web::PortletPreferences or Web::PortletProperties instance.
  #  - pl          only check public or private layouts? defaults to both. ( nil | :public | :private )
  #
  def select_layouts_with(portlet,pl=nil)
    name = (portlet.is_a?(String) ? portlet : portlet.name)
    layouts = (
      if pl.nil?
        self.layouts
      elsif pl==:public
        self.public_layouts
      elsif pl==:private
        self.private_layouts
      else
        []
      end
    )

#     layouts.select{|l| l.settings.include?(portlet)}
    layouts.select do |l|
      l.portlets.map {|p| p.name==name}.any?
    end
  end

  # Does any of the layouts include this portlet? See #select_layouts_with
  def layouts_include?(portlet,pl=nil)
    self.select_layouts_with(portlet,pl).any?
  end

  # Helper to collect all tags that are related to this Group through Tag::Asset.
  def assets_tags
    self.assets.collect(&:tags).flatten.uniq
  end

  # Unless the tagged_content portlet is found in any of the groups layouts,
  # a new layout (public by default) is created with this portlet.
  #
  # See #select_layouts_with
  #
  # The asset viewer is different in different versions of the Liferay database.
  #  - 5.1.x: tagged_content
  #  - 5.2.x: asset_publisher
  def asset_viewer_portlet(params={})
    if Lportal::Schema.buildnumber < 5200
      logger.debug 'Using lportal schema 5.1.x'
      params[:name] ||= 'tagged_content'
    else
      logger.debug 'Using lportal schema 5.2.x'
      params[:name] ||= 'asset_publisher'
    end

    portlet = Web::Portlet.find_by_name(params[:name])
    raise ('Portlet ”%s” not found! Check that Caterpillar migrations are up-to-date' % params[:name]) unless portlet

    # this says: if <tt>pl</tt> is undefined, use public layout, else use what <tt>pl</tt> specified.
    params.update(:privatelayout => ( params[:privatelayout].nil? ? false : params[:privatelayout] ))

    pl = (params[:privatelayout] ? :private : :public) # a small "type conversion"
    layouts = self.select_layouts_with(portlet,pl)

    ### Use existing layout and portlet
    if layouts.any?
      logger.debug 'Layout with tagged_content portlet found..'
      layout = layouts.first
      layout.portlets.each do |p|
        return p if p.name==portlet.name
      end
    end

    ### Create Layout + PortletPreferences
    logger.info 'Creating new Layout ”%s” in Group %s (%i)' % [params[:name],self.name,self.id]

    params[:friendlyurl] ||= ('/%s' % params[:name])
    params.update(:group => self)

    layout = Web::Layout.create(params)
    logger.debug 'Layout ready'
    logger.debug layout.inspect

    layout.settings = Web::Typesettings.new
    layout.columns=1

    # add the portlet to the layout
    layout.<<( portlet )
    layout.save

    portlet.reload
    logger.debug portlet.preferences.inspect
    return portlet.preferences
  end

end
